tests/apollo-fetch.ts

```diff
@@ -0,0 +1,496 @@
+import * as chai from 'chai';
+import * as chaiAsPromised from 'chai-as-promised';
+import { isEqual } from 'lodash';
+import gql from 'graphql-tag';
+import { print } from 'graphql';
+import * as fetchMock from 'fetch-mock';
+import {
+  createApolloFetch,
+} from '../src/apollo-fetch';
+import {
+  RequestAndOptions,
+  ResponseAndOptions,
+} from '../src/types';
+
+chai.use(chaiAsPromised);
+
+const { assert, expect } = chai;
+
+const sampleQuery = gql`
+query SampleQuery {
+  stub{
+    id
+  }
+}
+`;
+
+describe('apollo-fetch', () => {
+  const postData = {hello: 'world', method: 'POST'};
+  const data = JSON.stringify({data: { hello: 'world' }});
+  const unparsableData = 'raw string';
+  const mockError = { throws: new TypeError('mock me') };
+  const swapiUrl = 'http://graphql-swapi.test/';
+  const missingUrl = 'http://does-not-exist.test/';
+
+  const unauthorizedUrl = 'http://unauthorized.test/';
+  const serviceUnavailableUrl = 'http://service-unavailable.test/';
+
+  const simpleQueryWithNoVars = gql`
+    query people {
+      allPeople(first: 1) {
+        people {
+          name
+        }
+      }
+    }
+  `;
+
+  const simpleQueryWithVar = gql`
+    query people($personNum: Int!) {
+      allPeople(first: $personNum) {
+        people {
+          name
+        }
+      }
+    }
+  `;
+
+  const simpleResult = {
+    data: {
+      allPeople: {
+        people: [
+          {
+            name: 'Luke Skywalker',
+          },
+        ],
+      },
+    },
+  };
+
+  const complexQueryWithTwoVars = gql`
+    query people($personNum: Int!, $filmNum: Int!) {
+      allPeople(first: $personNum) {
+        people {
+          name
+          filmConnection(first: $filmNum) {
+            edges {
+              node {
+                id
+              }
+            }
+          }
+        }
+      }
+    }
+  `;
+
+  const complexResult = {
+    data: {
+      allPeople: {
+        people: [
+          {
+            name: 'Luke Skywalker',
+            filmConnection: {
+              edges: [
+                {
+                  node: {
+                    id: 'ZmlsbXM6MQ==',
+                  },
+                },
+              ],
+            },
+          },
+        ],
+      },
+    },
+  };
+
+  before(() => {
+
+    fetchMock.post('/graphql', data);
+    fetchMock.post('/raw', unparsableData);
+    fetchMock.post('begin:data', postData);
+    fetchMock.post('begin:error', mockError);
+    fetchMock.post('test', data);
+
+    fetchMock.post(swapiUrl, (url, opts) => {
```

Comment:
I think you probably want to write a utility function for easily mocking a server. There's an example of that in Apollo Client with mockNetworkInterface.